### ISSUE TITLE
Proposal to add `SecretSeed` trait(?) into `curve25519-dalek` crate as public service util.

### ISSUE THREAD

**Issue by SergeStrashko** [dalek-cryptography/curve25519-dalek#497]
Proposal:
Implement utility `Seed` like following
```rust
pub trait SecretSeed {
    fn random() -> Scalar; // generates new random and does `clamp_bytes`
    fn from_bytes(&[u8]) -> Scalar; //does `clamp_bytes` on input and returns Scalar
}

fn clamp_bytes(&[u8]) -> Scalar;

```
- Repos [`ed25519-dalek`](https://github.com/dalek-cryptography/ed25519-dalek) and [`x25519-dalek`](https://github.com/dalek-cryptography/x25519-dalek) can use this trait to obtain always compliant secret bytes. 
- It will (*as far as I can see - should*) allow to drop dependency on `rand` and `rand-core` in those repos.

Rationale:
1. https://github.com/dalek-cryptography/ed25519-dalek/blob/7529d65506147b6cb24ca6d8f4fc062cac33b395/src/public.rs#L150
```rust
    fn mangle_scalar_bits_and_multiply_by_basepoint_to_produce_public_key(
        bits: &mut [u8; 32],
    ) -> PublicKey {
        bits[0] &= 248;
        bits[31] &= 127;
        bits[31] |= 64;

        let point = &Scalar::from_bits(*bits) * &constants::ED25519_BASEPOINT_TABLE;
        let compressed = point.compress();

        PublicKey(compressed, point)
    }
```
2. https://github.com/dalek-cryptography/ed25519-dalek/blob/7529d65506147b6cb24ca6d8f4fc062cac33b395/src/secret.rs#L286 
```rust
    fn from(secret_key: &'a SecretKey) -> ExpandedSecretKey {
        let mut h: Sha512 = Sha512::default();
        let mut hash:  [u8; 64] = [0u8; 64];
        let mut lower: [u8; 32] = [0u8; 32];
        let mut upper: [u8; 32] = [0u8; 32];

        h.update(secret_key.as_bytes());
        hash.copy_from_slice(h.finalize().as_slice());

        lower.copy_from_slice(&hash[00..32]);
        upper.copy_from_slice(&hash[32..64]);

        lower[0]  &= 248;
        lower[31] &=  63;
        lower[31] |=  64;

        ExpandedSecretKey{ key: Scalar::from_bits(lower), nonce: upper, }
    }
```
3. https://github.com/dalek-cryptography/x25519-dalek/blob/53bb1a3989874ffe526a90fd225b5cb1c18f06c3/src/x25519.rs#L277
```rust
fn clamp_scalar(mut scalar: [u8; 32]) -> Scalar {
    scalar[0] &= 248;
    scalar[31] &= 127;
    scalar[31] |= 64;

    Scalar::from_bits(scalar)
}
```
4. https://github.com/dalek-cryptography/curve25519-dalek/blob/8c2f545d91f909fb7e6d2ec467c2218532790b8e/src/scalar.rs#L273
```rust
    /// Construct a `Scalar` from the low 255 bits of a 256-bit integer.
    ///
    /// This function is intended for applications like X25519 which
    /// require specific bit-patterns when performing scalar
    /// multiplication.
    pub const fn from_bits(bytes: [u8; 32]) -> Scalar {
        let mut s = Scalar { bytes };
        // Ensure that s < 2^255 by masking the high bit
        s.bytes[31] &= 0b0111_1111;

        s
    }
```
So both repos [`ed25519-dalek`](https://github.com/dalek-cryptography/ed25519-dalek) and [`x25519-dalek`](https://github.com/dalek-cryptography/x25519-dalek) require same functionality to make incoming random `[u8; 32]` bytes to be valid `secret key`. And as follows from(4) `Scalar` always makes sure that highest bit is dropped.

P.S. I think `X & 127 | 64 == X & 63 | 64`

P.P.S If it is `go for it!` I can create PRs implementing it.

**Comment by tarcieri** [comment]
Colloquially this set/clear bits to produce a valid scalar from an input bytestring procedure (which `ed25519-dalek` calls `mangle_scalar_bits`) has come to be referred to as "clamping".

I agree it'd be nice to have a first-class API for it, as the code for it is indeed duplicated both between and repeatedly inside of `ed25519-dalek` and `x25519-dalek`.

I think for starters you should just focus on a "clamping" API alone and leave out RNG changes for now. As a strawman:

```rust
impl Scalar {
    fn clamp(mut bytes: [u8; 32]) -> Self {
        bytes[0] &= 0b11111000;
        bytes[31] &= 0b111111;
        bytes[31] |= 0b1000000;
        Self::from_bits(bytes)
    }
}

**Comment by rozbb** [comment]
Agreed. Even thought it only appears in schemes built atop Curve25519, a `Scalar::clamp` function would probably be appropriate to include here.

**Comment by SergeStrashko** [comment]
#498

**Comment by rozbb** [comment]
Closed by #498
